py/simple_reporter.py

```diff
@@ -0,0 +1,327 @@
+    f.write('segment_id,next_segment_id,duration,count,length,queue_length,minimum_timestamp,maximum_timestamp,source,vehicle_type' + os.linesep)
+    for s in segments:
+      f.write(s)
+    thread_local.client.put_object(
+      Bucket=bucket,
+      Body=f.getvalue(),
+      Key='/'.join(file_name.split(os.sep)[1:]) + '/' + hashlib.sha1(file_name).hexdigest())  
+
+def report_tiles(match_dir, bucket, privacy, threads):
+  #download and parse them into proper files
+  logger.info('Reporting anonymised time tiles')
+  pool = ThreadPool(threads, local_session)
+  total = 0.0
+  for root, dirs, files in os.walk(match_dir):
+    for file_name in files:
+      pool.add_task(report, root + os.sep + file_name, bucket, privacy)
+      total += 1
+
+  #monitor progress
+  progress = -1
+  while pool.queue_size() > 0:
+   p = int((1.0 - (pool.queue_size() / total)) * 20) * 5
+   if p > progress:
+     logger.info('Reporting tiles %d%%' % p)
+     progress = p
+   time.sleep(1)
+  pool.join()
+  if progress != 100:
+    logger.info('Reporting tiles 100%')
+  logger.info('Done reporting tiles')
+  return dest_dir
+
+if __name__ == '__main__':
+  #build args
+  parser = argparse.ArgumentParser()
+  parser.add_argument('--src-bucket', type=str, help='Bucket where to get the input trace data from', required=True)
+  parser.add_argument('--src-prefix', type=str, help='Bucket prefix for getting source data', required=True)
+  parser.add_argument('--src-key-regex', type=str, help='Bucket key regex for getting source data', default='.*')
+  parser.add_argument('--src-keyer', type=str, help='A lambda used to extract the key from a given message in the input', default='lambda l: l.split("|")[1]')
+  parser.add_argument('--src-valuer', type=str, help='A lambda used to extract the time, lat, lon, accuracy from a given message in the input', default='lambda l: functools.partial(lambda c: [c[0], c[9], c[10], c[5] ], l.split("|"))()')
+  parser.add_argument('--src-time-pattern', type=str, help='A string used to extract epoch seconds from a time string', default='%Y-%m-%d %H:%M:%S')
+  parser.add_argument('--match-config', type=str, help='A file containing the config for the map matcher', required=True)
+  parser.add_argument('--quantisation', type=int, help='How large are the buckets to make tiles for. They should always be an hour (3600 seconds)', default=3600)
+  parser.add_argument('--privacy', type=int, help='How many readings of a given segment pair must appear before it being reported', default=2)
+  parser.add_argument('--source-id', type=str, help='A simple string to identify where these readings came from', default='smpl_rprt')
+  parser.add_argument('--dest-bucket', type=str, help='Bucket where we want to put the reporter output', required=True)
+  parser.add_argument('--concurrency', type=int, help='Number of threads to use when doing various stages of processing', default=multiprocessing.cpu_count())
+  args = parser.parse_args()
+
```

Comment:
3 main steps:

* spawn threads to download source trace data from s3, each thread will gunzip the file and parse out individual lines and forward them to a file for just that vehicle for that line. the final result is one file for every uniq vehicle (we'll need a decent size inode table for large datasets)
* spawn threads to take each file of per vehicle trace, do the match to the osmlr segments and then store the result to time tiles (what the java reporter outputs)
* spawn threads to take each time tile and remove the observations that dont meet the privacy parameter, then upload that file to s3

details of each can be found above